File: src/main/kotlin/org/kravbank/java/api/ProductResource.java
package org.kravbank.java.api;


import org.kravbank.java.model.Product;

import javax.transaction.Transactional;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.net.URI;
import java.util.List;

@Path("/products")

//@Produces(APPLICATION_JSON)
//@Consumes(APPLICATION_JSON)

public class ProductResource {
    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response getAll() {
        List<Product> products = Product.listAll();
        return Response.ok(products).build();
    }

    @GET
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getOne(@PathParam("id") Long id) {
        return Product.findByIdOptional(id)
                .map(product -> Response.ok(product).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }

    @POST
    @Transactional
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Response create(Product product) {
        Product.persist((product));
        if (product.isPersistent()) {
            return Response.created(URI.create("/products" + product.id)).build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }

    @GET
    @Path ("product/{title}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getByTitle(@PathParam("title") String title) {
        List <Product> products = Product.list("SELECT p from Product p WHERE p.title = ?1 ORDER BY id DESC", title);
        return Response.ok(products).build();
    }

    @DELETE
    @Path("{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Transactional
    public Response deleteById(@PathParam("id") Long id){
        boolean deleted = Product.deleteById(id);
        if (deleted){
            return Response.noContent().build();
        }
        return Response.status(Response.Status.BAD_REQUEST).build();
    }
}

File: src/main/kotlin/org/kravbank/api/ProductResource.kt
package org.kravbank.api

import org.kravbank.utils.form.product.ProductForm
import org.kravbank.utils.form.product.ProductFormUpdate
import org.kravbank.service.ProductService
import org.kravbank.service.ProjectService
import javax.enterprise.context.RequestScoped
import javax.transaction.Transactional
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

//@Tags(value = [Tag(name = "Read products", description = "Read uploaded products.")])
@Path("/api/v1/projects/{projectRef}/products")
@RequestScoped
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
//@SecurityScheme(securitySchemeName = "jwt", type = SecuritySchemeType.HTTP, scheme = "Bearer", bearerFormat = "JWT")
//@Authenticated
class ProductResource(val productService: ProductService, val projectService: ProjectService) {
    //GET PROJECT
    @GET
    @Path("/{productref}")
    fun getProduct(
        @PathParam("productref") productref: String,
        @PathParam("projectRef") projectRef: String,
    ): Response =
        productService.getProductByRefFromService(projectRef, productref)

    //LIST PRODUCTS
    @GET
    fun listProducts(@PathParam("projectRef") projectRef: String): Response =
        productService.list(projectRef)

    //CREATE PRODUCT
    @Transactional
    @POST
    fun createProduct(@PathParam("projectRef") projectRef: String, product: ProductForm): Response =
        productService.createProductFromService(projectRef, product)


    //DELETE PRODUCT
    @DELETE
    @Path("/{productref}")
    @Transactional
    fun deleteProdudctById(
        @PathParam("projectRef") projectRef: String,
        @PathParam("productref") productref: String
    ): Response =
        productService.deleteProductFromService(projectRef, productref)

    //UPDATE PRODUCT
    @PUT
    @Path("{productref}")
    @Transactional
    fun updateProduct(
        @PathParam("projectRef") projectRef: String,
        @PathParam("productref") productref: String,
        product: ProductFormUpdate
    ): Response =
        productService.updateProductFromService(projectRef, productref, product)
}

